src/merkle/partial_mt/mod.rs

```diff
@@ -0,0 +1,365 @@
+        // NoLeafForIndex("There is no leaf for provided index"). But it will be used almost never.
+        Err(MerkleError::NodeNotInSet(node_index))
+    }
+
+    /// Returns a value of the leaf at the specified NodeIndex.
+    ///
+    /// # Errors
+    /// Returns an error if the NodeIndex is not contained in the leaves set.
+    pub fn get_leaf(&self, index: NodeIndex) -> Result<Word, MerkleError> {
+        if !self.leaves.contains(&index) {
+            // This error not really suitable in this situation, should I create a new error?
+            Err(MerkleError::InvalidIndex {
+                depth: index.depth(),
+                value: index.value(),
+            })
+        } else {
+            self.nodes
+                .get(&index)
+                .ok_or(MerkleError::NodeNotInSet(index))
+                .map(|hash| **hash)
+        }
+    }
+
+    /// Returns a map of paths from every leaf to the root.
+    pub fn paths(&self) -> Result<BTreeMap<&NodeIndex, MerklePath>, MerkleError> {
+        let mut paths = BTreeMap::new();
+        for leaf_index in self.leaves.iter() {
+            let index = *leaf_index;
+            paths.insert(leaf_index, self.get_path(index)?);
+        }
+        Ok(paths)
+    }
+
+    /// Returns a Merkle path from the node at the specified index to the root.
+    ///
+    /// The node itself is not included in the path.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// - the specified index has depth set to 0 or the depth is greater than the depth of this
+    /// Merkle tree.
+    /// - the specified index is not contained in the nodes map.
+    pub fn get_path(&self, mut index: NodeIndex) -> Result<MerklePath, MerkleError> {
+        if index.is_root() {
+            return Err(MerkleError::DepthTooSmall(index.depth()));
+        } else if index.depth() > self.max_depth() {
+            return Err(MerkleError::DepthTooBig(index.depth() as u64));
+        }
+
+        if !self.nodes.contains_key(&index) {
+            return Err(MerkleError::NodeNotInSet(index));
+        }
+
+        let mut path = Vec::new();
+        for _ in 0..index.depth() {
+            let sibling_index = Self::get_sibling_index(&index)?;
+            index.move_up();
+            let sibling_hash =
+                self.nodes.get(&sibling_index).cloned().unwrap_or(Self::EMPTY_DIGEST);
+            path.push(Word::from(sibling_hash));
+        }
+        Ok(MerklePath::new(path))
+    }
+
+    // ITERATORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns an iterator over the leaves of this [PartialMerkleTree].
+    pub fn leaves(&self) -> impl Iterator<Item = (NodeIndex, &Word)> {
+        self.nodes
+            .iter()
+            .filter(|(index, _)| self.leaves.contains(index))
+            .map(|(index, hash)| (*index, &(**hash)))
```

Comment:
Wouldn't it be easier to iterate over `leaves` and for each leaf index look up the relevant node?